프로젝트 관리 앱 [STEP 2] 민쏜

minsson opened the pull request:
@GREENOVER 

그린, 안녕하세요! 😁
민쏜입니다!

Step 2 PR 보냅니다!
그린의 모든 조언이 소중합니다..! 
가감없는 피드백 부탁 드립니다..!! 🔥

감사합니다! ☺️

---

# 1. 화면 구현 📱

### 1-1. 메인 화면
| Light Mode | Dark Mode|
|:-----------:|:----------:|
| <img src=https://i.imgur.com/JKLG5TQ.png width=100%> | <img src=https://i.imgur.com/DjdkVAO.png width=100%> |

### 1-2. Task 생성 화면
| Light Mode | Dark Mode|
|:-----------:|:----------:|
| <img src=https://i.imgur.com/TmJ4shC.png width=100%> | <img src=https://i.imgur.com/vM1DCbd.png width=100%> |

### 1-3. Task 수정 화면
| Light Mode | Dark Mode|
|:-----------:|:----------:|
| <img src=https://i.imgur.com/Mx93een.png width=100%> | <img src=https://i.imgur.com/rvq8bdz.png width=100%> |

### 1-4. Task 진행상태 변경 및 삭제
| Light Mode |
|:-----------:|
|![](https://i.imgur.com/kFPQc8k.gif)|

| Dark Mode |
|:-----------:|
|![](https://i.imgur.com/1oRNMZV.gif)|

# 2. 폴더 및 파일 구조 🌴

```
.
└── ProjectManager/
    ├── GoogleService-Info.plist
    ├── Resources/
    │   └── Assets.xcassets
    └── Sources/
        ├── ProjectManagerApp
        └── Common/
            ├── Enum/
            │   └── Status
            ├── Extensions/
            │   └── Date+Extensions
            ├── Models/
            │   ├── Task
            │   └── TasksDataSource
            ├── Views/
            │   ├── Common/
            │   │   ├── SquareButtonView
            │   │   └── AddingTaskButtonView
            │   ├── MainView/
            │   │   └── TaskDashboardView
            │   ├── TaskList/
            │   │   ├── TaskListHeaderView
            │   │   └── TaskListView
            │   ├── TaskCell/
            │   │   ├── TaskCellBackgroundView
            │   │   └── TaskCellView
            │   └── TaskCreating+TaskEditing/
            │       ├── Protocol/
            │       │   └── TaskWritableView
            │       ├── TaskCreatingView
            │       ├── TaskEditingView
            │       └── TaskDescriptionView
            └── Preview Content/
                └── Preview Assets
```



# 3. 고민한 점, 조언이 필요한 점 🧩

## 3-1. MV vs MVVM, Model과 View의 역할

처음에는 MVVM 아키텍처를 사용하려고 했는데, 그린 추천으로 아래와 같은 글을 읽고 SwiftUI에서 굳이 ViewModel이 필요하지 않다는 점에 대해 공감했습니다.
- [Stop using MVVM for SwiftUI](https://developer.apple.com/forums/thread/699003)
- [SwiftUI에서 MVVM 사용을 멈춰야 하는가?](https://green1229.tistory.com/267)

그래서 일단은 ViewModel 없이 Model과 View로 구성해봤습니다. 서비스 로직과 비즈니스 로직이라는 용어에 대해 생각해볼 때, 쉽게 생각해보면 뷰에는 서비스 로직을 넣으면 될 것 같고, SOT 역할을 하는 모델에는 비즈니스 로직을 넣으면 될 거라고 생각했습니다. 그런데 막상 구현을 하는 과정에서 역할 분리에서 명확하게 이해되지 않는 부분이 어느 정도 있었습니다. 특히 `뷰에서 이런 걸 해도 괜찮을까?`라고 고민되는 지점이 있었습니다.

>**🙋‍♂️ 질문**
>그린이 보시기에는 어떤지, 구조적으로 어떤 점을 개선하면 좋을지 가르쳐주시면 감사하겠습니다. 우선 제가 애매하게 느껴졌던 부분은 코드 리뷰 쪽에 먼저 코멘트를 달아놓겠습니다.

>**🙋‍♂️ 질문**
>앞으로 SwiftUI를 꾸준히 공부하고 사용하고 싶습니다. 앞으로 어떤 아키텍처를 더 공부해보면 좋을까요? 그린의 블로그에서도 TCA에 대한 이야기를 많이 다루고 있어, 한 번 공부해보고 싶은데 먼저 공부하거나 체득해야할 개념이 있을까 해서 방향성에 대해 여쭤봅니다.


## 3-2. 데이터 전달(Environment Object vs Observed Object)

Environment Object를 처음 공부할 때, 이를 싱글톤과 비슷하다고 설명하는 글이 많았습니다. 그래서 우선 가급적 사용하지 않아야겠다는 생각이 들었습니다. 처음 싱글톤 패턴을 배우고 사용했을 때 그게 너무 편해 남발한 경험이 있고, 그로 인해 싱글톤 패턴의 여러 단점 중 하나인 테스트가 어려워지는 문제를 겪어본 적이 있기 때문입니다.

그런데 Environment Object에 대해 공부를 하다 보니, 마땅한 단점이 보이지 않았습니다. 

아래는 단점을 찾아보려고 했을 때 들었던 의문과, 그에 대한 셀프 반박입니다.
- Singleton에서의 단점과 같이, 어디에서나 쓰일 수 있다.
    - 뷰 안에서 @EnvironmentObject 프로퍼티 래퍼를 사용해 받아야 쓸 수 있으므로, 코드에서 갑자기 보이는 Singleton과는 차이가 있다는 생각이 들었습니다.
- 하나의 객체가 메모리에 생성되어, 사용하지 않는 뷰에서도 여전히 메모리를 점유하고 있다.
    - 객체의 크기가 상식적으로 몇 킬로바이트도 되지 않을 것 같습니다. 현 시대의 디바이스 스펙으로 걱정할 이유가 없다는 생각이 들었습니다.
- 테스트하기가 어렵다.
    - 아직 테스트를 구현해보지 않았지만, 이 부분은 검색해보니 문제될 게 없어보였습니다. (이건 직접 테스트를 진행하면서 다시 한번 알아보려고 합니다.)

>**🙋‍♂️ 질문**
>저는 Environment Object의 단점을 아직 알아내지 못했는데, 혹시 그린은 어떻게 생각하시나요?


## 3-3. 폴더 및 파일 구조
처음 SwiftUI로 프로젝트를 진행하다보니 파일 구조를 어떻게 해야할지 고민이 되었습니다. 일단은 Feature 단위로 묶는다고 생각하고 정리해봤는데 그린은 어떻게 생각하시는지 궁금합니다.

```
.
└── ProjectManager/
    ├── GoogleService-Info.plist
    ├── Resources/
    │   └── Assets.xcassets
    └── Sources/
        ├── ProjectManagerApp
        └── Common/
            ├── Enum/
            │   └── Status
            ├── Extensions/
            │   └── Date+Extensions
            ├── Models/
            │   ├── Task
            │   └── TasksDataSource
            ├── Views/
            │   ├── Common/
            │   │   ├── SquareButtonView
            │   │   └── AddingTaskButtonView
            │   ├── MainView/
            │   │   └── TaskDashboardView
            │   ├── TaskList/
            │   │   ├── TaskListHeaderView
            │   │   └── TaskListView
            │   ├── TaskCell/
            │   │   ├── TaskCellBackgroundView
            │   │   └── TaskCellView
            │   └── TaskCreating+TaskEditing/
            │       ├── Protocol/
            │       │   └── TaskWritableView
            │       ├── TaskCreatingView
            │       ├── TaskEditingView
            │       └── TaskDescriptionView
            └── Preview Content/
                └── Preview Assets
```



## 3-4. SSOT(Single Source of Truth)
프로젝트의 `Models` 그룹을 보면 `Task`와 `TasksDataSource`가 있습니다.
`Task`는 모델 객체 그 자체고, `TasksDataSource`는 `Single Source of Truth`의 역할을 하고 있습니다.

>**🙋‍♂️ 질문**
>SSOT 역할을 하는 `TasksDataSource`의 네이밍과 역할 분리가 잘 되어 있다고 볼 수 있을까요?


## 3-5. Task 데이터 관리
`TasksDataSource`를 보면, 아래와 같이 각 Task가 담길 어레이를 진행상태(status) 별로 분류하고 있습니다. 

```swift
final class TasksDataSource: ObservableObject {
    @Published var todoTasks: [Task] = []
    @Published var doingTasks: [Task] = []
    @Published var doneTasks: [Task] = []
}
```

그러다보니 모든 비즈니스 로직에서 `switch-case`를 사용해야 한다는 불편함이 있었습니다. 예를 들어, 아래의 메서드처럼요.

```swift
func transfer(selectedTask: Task, to newCategory: Status) {
        deleteOriginalTask(equivalentTo: selectedTask)
        
    switch newCategory {
    case .todo:
        let newTask = newTask(copiedFrom: selectedTask, withNewStatus: .todo)
        todoTasks.append(newTask)
    case .doing:
        let newTask = newTask(copiedFrom: selectedTask, withNewStatus: .doing)
        doingTasks.append(newTask)
    case .done:
        let newTask = newTask(copiedFrom: selectedTask, withNewStatus: .done)
        doneTasks.append(newTask)
    }
}
```

>**🙋‍♂️ 질문**
>처음에는 아래와 같은 코드를 작성했습니다. (TaskDashboardViewModel의 네이밍을 TasksDataSource로 수정하기 전의 코드입니다)
>```swift
>class TaskDashboardViewModel: ObservableObject {
>    
>    @Published var tasks: [Task] = []
>    
>    var todo: [Task] {
>        tasks.filter { $0.status == .todo }
>    }
>    
>    var doing: [Task] {
>        tasks.filter { $0.status == .doing }
>    }
>    
>    var done: [Task] {
>        tasks.filter { $0.status == .doing }
>    }
>}
>```
> 그러다가 아래 9b2246cf1ca45e3004d8e4a1ce749411f9fdb765 번 커밋의 커밋 메시지와 같은 이유로 수정해 현재 코드가 되었습니다. 그린은 두 방안의 장단점에 대해 어떻게 생각하시는지 궁금합니다. 또, 그린이라면 이런 방법 외에 혹시 어떻게 하셨을지도 궁금합니다!

> 커밋 메시지 (9b2246cf1ca45e3004d8e4a1ce749411f9fdb765)
>- "TaskDashboardViewModel"
>   - tasks라는 하나의 어레이 타입 변수로 관리하던 Task들을 세 개로 분리
>   - 기존의 연산프로퍼티를 이용해 status별로 TaskListView에서 띄우는 경우, List에서 보는 indexSet과 실제 tasks 변수의 index를 연결할 수 없어 선택한 차선
>       - 이 문제는 List에서 id를 조회하는 방식으로 해결 가능하다는 것을 이후에 알게 되었음
>       - 하지만 task들을 status 별로 따로 관리하는 구조가 더 직관적이고, 매번 연산프로퍼티의 filter를 사용할 일이 없어 효율적이라고 생각해 현재 수정방안을 유지하기로 함
>   - 각 어레이 안의 데이터의 경우 임시로 넣어놓은 데이터로 추후 삭제할 예정


## 3-6. View를 만드는 방법
처음에는 모든 뷰를 구조체로 만들어줘야 한다고 생각했습니다. 그런데 그 외에도 메서드나 연산 프로퍼티로 구현하는 방법도 있었습니다.

우선 커스텀 버튼 등 다양한 곳에서 사용할 수 있는 뷰는 구조체로 구현하는 게 좋을 것 같다고 생각했습니다.

그런데 특정 뷰 내부에서만 쓰일 뷰를 구현하면서 고민이 시작되었습니다.

- Nested struct로 구현
- 인스턴스 메서드로 구현 (저는 이 방법을 선택했습니다. 구현하기에 가장 간단하게 느껴졌기 때문입니다.)
- 연산 프로퍼티로 구현

>**🙋‍♂️ 질문**
>각 방법에 대한 장단점과 그린의 선택 기준이 궁금합니다! 또, 이 방법 외에도 다른 방법이 있는지 궁금합니다.


## 3-7. Task 생성 뷰와 편집 뷰
Task를 생성하고 수정하기 위한 sheet를 구현할 때 이 부분에서 고민을 많이 했습니다. 처음에는 하나의 뷰에서 분기처리를 통해 구현하려고 했는데, 바인딩 프로퍼티 래퍼가 복잡해지고, 신규 Task인지 기존 Task의 수정인지에 따른 `수정` 버튼의 유무와 기능, 이를 추적하기 위한 프로퍼티 등등... 프로퍼티도, 이니셜라이저도, 분기처리도 모두 복잡해져 득보다 실이 많다는 생각이 들었습니다.

하지만 뷰의 전체적인 모양 자체는 동일해 중복된 코드가 일부 존재하는 게 적절하지 않다고 생각했고, `TaskWritableView`라는 프로토콜과 기본 구현을 통해 중복 코드를 없애보았습니다.

>**🙋‍♂️ 질문**
>나름대로 해봤지만, 아직 더 개선할 수 있을 것 같은데, 어떻게 개선하면 좋을지 궁금합니다.

>**🙋‍♂️ 질문**
>프로젝트 자체와는 관련이 없는 번외 질문인데, 그린처럼 현업에 계신 분들의 코드를 보면 프로토콜의 프로토콜의 프로토콜...과 같이 추상화 단계가 무척 깊은 걸 볼 수 있는데요. 어떤 성장 과정을 통해 그런 걸 자연스럽게 할 수 있게 되신 건지 궁금합니다. (최근에 디자인 패턴을 공부한 적이 있는데, 혹시 디자인 패턴을 공부하시면서 프로토콜을 더욱 잘 활용하게 되신 걸까...? 라는 생각이 들기도 했습니다 😅)


## 3-8. 마감기한 초과 판정 로직
TaskCellView에서 Task를 보여줄 때, 마감기한이 지난 경우 날짜를 빨간색으로 표시해주고 있습니다.

처음에는 단순히 `dueDate < Date.now`이면 빨갛게 표시하면 될 것 같다고 생각했는데, 다시 생각해보니 이런 대소 비교는 `TimeInterval`, 즉, `Double`을 통해 이루어지니, `마감일과 오늘이 같은 날짜라도 시간 차이에 의해 빨갛게 표시되는 문제`가 생길 수 있다고 생각했습니다.

자주 구현할 것 같은 기능이라 Foundation Framework에 관련 메서드가 있을 거라고 생각했는데, 여러 방면으로 검색해봐도 찾지 못했습니다.

대신 블로그 등 다른 분들이 유사한 케이스에서 구현하신 걸 보면 날짜를 컴포넌트처럼 일, 시, 분, 초 등으로 나누어 구현하는 경우가 많았습니다.

>**🙋‍♂️ 질문**
아래와 같이 구현하면 코드를 최소화할 수 있다고 생각했는데, 혹시 이런 식으로 구현하는 것도 괜찮을까요?
>```swift
>Text(task.dueDate, formatter: Date.formatter)
>    .foregroundColor(isOver(task.dueDate) ? .red : .primary)
>```
>
>```swift
>private extension TaskCellView {
>    func isOver(_ dueDate: Date) -> Bool {
>        if dueDate < Date.now && !Calendar.current.isDateInToday(dueDate) {
>            return true
>        }
>        
>        return false
>    }
>}
>```

minsson committed:
chore: 프로젝트 폴더 및 파일 구조 정리

minsson committed:
feat: "Task" 모델 타입 구현

- "Status" Task의 할일, 진행중, 완료 여부를 나타낼 타입 구현

minsson committed:
feat: "TaskDashboardViewModel" Task 로직을 위한 뷰모델 구현

- Task들을 보관하기 위한 저장소의 역할
- Task의 status별로 구분하여 조회할 수 있게 해주는 역할

minsson committed:
feat: "TaskCellView" 하나의 Task를 담기 위한 뷰 구현

- "Date+Extensions" Date 타입을 원하는 포멧의 String으로 간편하게 변환하기 위한 메서드 구현

minsson committed:
feat: "TaskListView" Task Cell들을 담기 위한 뷰 구현

- "TaskCellBackgroundView" TaskCellView를 올려 놓는 커스텀한 백그라운드 뷰
- "TaskDashboardViewModel" 구현 중 참고를 위한 임시 더미데이터 추가
- "TaskCellView" 텍스트의 intrinsic size가 아닌, 텍스트가 올라가 있는 뷰 전체에서 왼쪽 정렬되도록 수정

minsson committed:
refactor: "TaskCellView" 셀의 백그라운드를 결정하는 뷰 추가

- TaskCellView의 백그라운드를 결정하는 TaskCellBackgroundView가 TaskListView에서 호출되는 건 부자연스러우므로 위치 이동
- 백그라운드와 텍스트들 간의 적절한 간격을 위해 padding 부여

- "TaskListView"
    - 호출이 부자연스러웠던 TaskCellBackgroundView 제거
    - TaskCellView에서 padding이 추가되어 불필요해진 frame modifier를 제거

minsson committed:
feat: "TaskListView" 주입 받은 status에 따라 다른 뷰가 되도록 구현

- "TaskDashboardViewModel" 오류 수정

minsson committed:
feat: "TaskListView" 리스트에서 Task를 스와이프해 삭제하는 액션 추가

- .listStyle(.plain) 모디파이어를 추가하여 List에서 기본으로 제공하는 백그라운드 색상 제거

- "TaskDashboardViewModel"
    - tasks라는 하나의 어레이 타입 변수로 관리하던 Task들을 세 개로 분리
    - 기존의 연산프로퍼티를 이용해 status별로 TaskListView에서 띄우는 경우, List에서 보는 indexSet과 실제 tasks 변수의 index를 연결할 수 없어 선택한 차선
        - 이 문제는 List에서 id를 조회하는 방식으로 해결 가능하다는 것을 이후에 알게 되었음
        - 하지만 task들을 status 별로 따로 관리하는 구조가 더 직관적이고, 매번 연산프로퍼티의 filter를 사용할 일이 없어 효율적이라고 생각해 현재 수정방안을 유지하기로 함
    - 각 어레이 안의 데이터의 경우 임시로 넣어놓은 데이터로 추후 삭제할 예정

minsson committed:
feat: "TaskListHeaderView" 각 List의 타이틀을 보여줄 헤더뷰 구현

- TaskListView로부터 보여줄 status를 전달 받고, status의 description을 띄우는 방식

- "Status" TaskListHeaderView에서 사용하기 위한description 연산프로퍼티 구현

- "TaskListView" TaskListHeaderView와 task의 List를 HStack으로 함께 보여주도록 변경

minsson committed:
feat: "AddingTaskButtonView" 할 일 추가를 위한 플로팅 버튼 구현

- "TaskDashboardView" 내부에 적용

minsson committed:
feat: "SquareButtonView" 재사용을 위한 커스텀 버튼 타입 구현

- 이름, 색상, 클로저를 받을 수 있는 버튼

minsson committed:
feat: "TaskEditingView" Task 작성, 수정, 저장을 위한 시트뷰 구현

- 시트뷰의 높이를 조절할 방법을 찾지 못해, 우선 구현 완료 후 추후 높이 조절을 시도해볼 예정

minsson committed:
fix: "TaskCellView" Cell에 표시할 항목별 최대 Line Limit 설정

minsson committed:
feat: "TaskEditingView" 수정시 수정버튼을 눌러야 수정되는 기능 구현

- isNewTask 변수를 통해 초기화시 Bool 타입 아규먼트를 전달받아 신규 작성인지 기존 내용 수정인지 구분
    - false를 받으면 기존 내용 수정이므로 수정 버튼을 눌러야 TextField 내의 내용을 변경할 수 있도록 함
    - true를 받으면 신규 작성이므로 즉시 작성이 가능하도록 함
- 현재는 무조건 신규 작성으로 진입하며, 기존 작성되어 있던 Task를 탭하여 진입하는 경우는 추가 예정

minsson committed:
refactor: "TasksDataSource" Task들의 SOT 역할을 할 객체 구현

- 필요한 뷰에 EnvironmentObject로 제공됨
- 기존의 "TaskDashboardViewModel"이라는 네이밍을 변경
    - MVVM 구조로 진행할 계획이던 프로젝트에서 ViewModel을 제거하고자 함

minsson committed:
feat: "TasksDataSource" 기존 항목 수정에 쓸 메서드 구현

- 기존 Task를 탭하여 수정 후 저장할 때 SOT의 원본 데이터를 새로 수정된 데이터로 변경하기 위함

minsson committed:
feat: "TaskListView" 기존 항목를 탭하여 수정, 저장하는 기능 구현

- 기존 항목 탭시 TaskEditingView sheet를 띄워, 수정 버튼을 누른 후 수정 및 저장하도록 구현
- 추후 UI 작업을 위한 프리뷰 재생성

- "TaskEditingView"
    - 새로운 항목 추가 및 기존 항목 수정을 위한 뷰에서 기존 항목을 수정하기 위한 뷰로 변환
        - 추가 뷰와 수정 뷰의 역할을 분리하고, UI 구성을 다르게 하기 위함

minsson committed:
feat: "TaskCreatingView" 객체 및 첫화면에서 할일을 생성하는 기능 구현

minsson committed:
feat: "TaskListView" 스와이프 액션을 활용하여 Task 삭제 기능 구현

minsson committed:
feat: "TaskListView" 스와이프로 각 Task의 진척도를 변경하는 기능 구현

- 각 Task 셀에 leading swipe 할 시, Task의 원래 status를 제외한 나머지 진척도 버튼을 표시함

- "TaskDataSource"
    - Task를 다른 status로 이동 시, 목적지 카테고리와 부합하는 newStatus가 포함된 Task 객체 생성 후 반환하는 메서드 구현
    - Task를 이동하고자 하는 status에 맞는 어레이로 보내주기 위한 메서드 구현

minsson committed:
feat: "TaskDescriptionView" Task 내용 작성용 텍스트에디터 구현

- 기본 TextEditor를 TaskCellView에서 사용할 경우 뒷배경이 하얀색이 되는 문제가 있었음
- TextEditor가 .onAppear 될 때 .clear 되도록 만들기 위해 이 타입을 구현함

minsson committed:
feat: "Date+Extensions" 날짜 표시 형식 한국어에 맞게 변경

- TaskCellView에서 영어 날짜 포멧으로 나오던 것을 한국 날짜 포멧으로 변경

minsson committed:
feat: "TaskCellView" 마감기한이 지난 Task의 날짜 색을 빨갛게 표시

- Task의 마감기한이 현재 일시에 도달하지 못했거나 OR 현재일 24시간 안에 포함된다면 아직 마감기한이 지나지 않은 것
    - 두번째 조건도 고려한 이유는 Date 타입에는 시간 정보 또한 포함되어, 마감일이 오늘인데도 빨간색으로 표시될 수 있기 때문
    - 따라서 마감기한이 오늘 중이기만 하면 두번째 조건을 만족하게 되어 빨간색으로 표시되지 않음

minsson committed:
refactor: "TaskWritableView" 할일 추가 및 수정 관련 뷰 기본 구현

- 기존의 "TaskCreatingView"와 "TaskEditingView"에서 중복으로 사용되던 뷰들을 프로토콜 확장 기본구현으로 정리

minsson committed:
feat: 다크모드에서의 사용자 경험 향상을 위한 일부 컬러 변경, 강조색 지정

minsson committed:
feat: Task 추가, 수정, 상태변경 시 사용자 경험 향상을 위한 애니메이션 구현

minsson committed:
style: 가독성 향상을 위한 일부 코드 위치 변경

- 일부 코드 접근제어 추가

minsson committed:
chore: 기능별 폴더 구조 및 파일 정리

minsson committed:
feat: 스와이프 액션 버튼에 사용할 Color Set 지정

- "TaskListView"
    - Color Set 적용
    - 불필요한 주석 제거

GREENOVER commented on ProjectManager/ProjectManager/Sources/Models/TasksDataSource.swift:
```diff
@@ -0,0 +1,104 @@
+        switch selectedTask.status {
+        case .todo:
+            let index = todoTasks.firstIndex { $0.id == selectedTask.id }
+                todoTasks.remove(at: index!)
```
index 강제 추출을 지양해보는건 어떨까요?🙋🏻
앱 크래쉬의 주요 원인이 됩니다..!

minsson replied:
구현 중 문제 발생시 바로 크래쉬가 나도록 강제언래핑을 사용했었는데, 이후 수정하지 않고 넘어갔었네요! 😅 
아래와 같이 if let 바인딩으로 언래핑하도록 수정했습니다.

```swift
        switch selectedTask.status {
        case .todo:
            let index = todoTasks.firstIndex { $0.id == selectedTask.id }
            
            if let index = index {
                todoTasks.remove(at: index)
            }
```

💾 40b140a62698f53e7c98e9ae60bf1edea84afcf8

GREENOVER commented on ProjectManager/ProjectManager/Sources/Models/TasksDataSource.swift:
```diff
@@ -0,0 +1,104 @@
+    }
+}
+
+private extension TasksDataSource {
```
private한 메서드와 internal 메서드의 extension으로 구분 좋네요!👍

GREENOVER commented on ProjectManager/ProjectManager/Sources/Views/Common/AddingTaskButtonView.swift:
```diff
@@ -0,0 +1,29 @@
+    var body: some View {
+        
+        Circle()
+            .overlay(alignment: .bottomTrailing) {
```
overlay와 ZStack은 어떤 차이가 있을까요?
overlay로 구현하신 이유가 궁금합니다🙋🏻
잘못된거 전혀 아닙니다!

minsson replied:
처음 구현할 때는 여러 튜토리얼을 참고했는데, 대부분 `ZStack`보다는 `overlay`를 먼저 알려주어, `overlay`가 좀 더 익숙했습니다. 그래서 처음 구현할 때는 `ZStack`을 고려하지 않고, 바로 `overlay`로 구현했습니다. 😅

두 개념 모두 뷰 위에 다른 뷰를 쌓아간다는 점에서는 비슷해보입니다. 

다만, 두 가지 차이점이 있어 보입니다.
1) `overlay`는 기존의 뷰를 기준으로 fit 되어 기존의 프레임에 갇혀있는 반면, `ZStack`은 쌓여있는 가장 큰 뷰를 기준으로 `frame`이 잡힌다는 점에서 차이가 있는 것 같습니다.
2) `ZStack`의 경우 `zIndex`로 쌓여있는 뷰들의 순서를 변경할 수 있지만, `overlay`한 것은 무조건 위에 덮이는 구조로 되어 있습니다.

이 점을 잘 고려하면 두 개념 모두 구현 방법은 다르지만 같은 결과물을 만들어낼 수 있을 것 같습니다.

제 생각은 위와 같은데, 혹시 둘 중 하나로 밖에 구현할 수 없는 상황이나, 아니면 일반적인 선택 기준 같은 게 있을까요? 🙋🏻

GREENOVER replied:
일반적인 선택기준은 민쏜이 차이를 잘 알고 말씀해주신것이 기준이 될것 같아요~!

GREENOVER commented on ProjectManager/ProjectManager/Sources/Views/MainView/TaskDashboardView.swift:
```diff
@@ -0,0 +1,45 @@
+    @EnvironmentObject private var tasksDataSource: TasksDataSource
+    
+    @State private var isShowingSheet = false
+    @State private var newTask = Task(title: "", description: "", dueDate: Date.now, status: .todo)
```
이 프로퍼티는 어디서 사용되나요?
실제로 사용이 되고 있는걸까요?🙋🏻

minsson replied:
사용하지 않는 프로퍼티이므로 제거했습니다! 😅

Step 2 구현 초기에 TaskDashboardView에서 sheet를 띄우면서 newTask를 주입해주는 방식으로 구현했었는데, newTask를 만들고 관리하는 역할을 TaskCreatingView로 분리하게 되면서 더 이상 사용하지 않게된 프로퍼티입니다. 

TaskCreatingView를 구현하면서 제거해야 했는데, 실수로 남겨 놓았네요 🥲

💾 0d26e9f2d8dfec509dc99922ff6614f7115638ea

GREENOVER commented on ProjectManager/ProjectManager/Sources/Views/MainView/TaskDashboardView.swift:
```diff
@@ -0,0 +1,45 @@
+                AddingTaskButtonView()
+                    .frame(width: 70, height: 70, alignment: .bottomTrailing)
+                    .padding(EdgeInsets(top: 0, leading: 0, bottom: 20, trailing: 40))
+            }
```
AddingTaskButtonView를 공통의 컴포넌트로 설계해주셨는데 이 뷰가 버튼 외에 다른 용도로 사용되지 않는다면
해당 컴포넌트 자체를 버튼으로 만들어 호출만 하면 어떨까요?🙋🏻

minsson replied:
해당 컴포넌트 자체를 버튼으로 만들어 호출하도록 변경했습니다! 

처음에는 버튼 종류는 추후 재사용될 여지가 있으므로 공통의 컴포넌트로 만들어보았는데, 말씀해주신 것처럼 다른 용도로 사용되지 않고, 추후에도 사용될 것으로 보이지 않아 TaskDashboardView의 private method로 만들었습니다!

💾 1faeb1b5fa5a29230da66e2762b804d420ba4ece

GREENOVER commented on ProjectManager/ProjectManager/Sources/Views/TaskList/TaskListHeaderView.swift:
```diff
@@ -0,0 +1,39 @@
+    var taskCount: Int
+    
+    var body: some View {
+        ZStack {
```
ZStack으로 구성해주신 이유는 이해가 되는데 더 간단히 HStack으로만 status와 count를 담아 적절한 Spacer를 통해 구현해줄 수 있을것 같습니다!🙋🏻
거의 그럴일은 없겠지만 현재 구조에서는 count의 수가 예를들어 300000이 오면 겹쳐 보일것 같습니다.

minsson replied:
taskCount를 임의로 수정한 결과, 그린이 말씀해주신 것처럼 ZStack이므로 겹쳐보이는 결과가 나옵니다.
![](https://i.imgur.com/Uqu1dyl.png)

이를 감안하고 ZStack로 구성한 이유는 아래와 같습니다.
1) status를 전체 width의 가운데에 놓고자 했습니다.
2) 현실적으로 Task의 개수가 매우 많아져 status를 침범하는 일은 없을 거라고 생각했습니다. (현실에서 수백만, 수천만 개의 Task를 수행해야 한다면... 🙀)

하지만 이런 식의 디자인 패턴(개발자의 디자인 패턴이 아닌, 디자이너들이 하는 디자인의 패턴)은 Task 관리 앱 외에도 얼마든지 있을 수 있고, 그럴 경우에는 그린이 조언해주신 것처럼 겹칠 여지가 충분하다는 생각이 듭니다! 😲
예를 들어, 제목 옆에 해쉬태그 같은 걸 표시하는 디자인이 있을 것 같네요!

그래서 겹치지 않게 만드는 코드를 작성해봤습니다.
```swift
            HStack {
                Spacer()

                Text(status.description)
                    .font(.largeTitle)
                    .fontWeight(.heavy)
                
                Spacer()

                Text(String(10000000000000000)) // taskCount의 자리에 임시로 임의의 수 할당
                    .foregroundColor(.white)
                    .padding()
                    .background(Color.accentColor)
                    .frame(height: 30, alignment: .trailing)
                    .cornerRadius(25)
            }
```
위 코드의 결과물은 아래와 같습니다.
![](https://i.imgur.com/ZbjqW6U.png)

여기까지는 좋은데, 아래와 같이 타이틀이 정가운데 오지 않고 `taskCount`의 `width`만큼 옆으로 밀리는 아쉬움이 있습니다. 겹치지 않을 때는 가급적 타이틀이 가운데에 있는 게 더 예쁠 것 같은데 말이죠..! (물론 이 부분은 제가 디자이너가 아니라 확신은 없지만요 😅)
![](https://i.imgur.com/m9cuPgt.png)
![](https://i.imgur.com/iaP0cad.png)

그래서 위 코드에서 status 좌측의 `Spacer()`에 `Spacer().frame(width: @@@)`을 추가로 주는 방법도 생각해보았는데, 딱히 의미가 있어보이지는 않았습니다.

혹시 그린이 생각하시는 더 좋은 방법이 있을까요? 🙋🏻

GREENOVER replied:
민쏜이 알아보신 차이 정확해요!
더 좋은 방법이라함은 지금 현재 민쏜이 알아보시고 제시해주신 방법에서는 불가능할것 같고,
일반적으로는 저정도로 많은 건을 표시해야되면 999+와 같이 기획에서 요구하는게 맞을거 같아요.
그래서 한번 이 부분도 생각해보면 좋을것 같아서 남겨보았어요!
정리하자면 ZStack까지 사용할 필요가 없이 HStack으로 표현해도 되지 않을까했고 그럴때 겹치지는 않지만 무수히 많아지면 타이틀을 가운데 위치시키지 못한다는점이 단점이고 ZStack에서는 겹치는게 단점일거 같아요.
그래서 둘다를 포용하는 방법으로는 기획을 수정하는것이 있을것 같습니다🙋🏻

GREENOVER commented on ProjectManager/ProjectManager/Sources/Views/TaskList/TaskListView.swift:
```diff
@@ -0,0 +1,190 @@
+            
+            List {
+                ForEach(tasks) { task in
+                    ZStack {
```
여기서 ZStack이 하는 역할이 뭔가요?🙋🏻

minsson replied:
구현 초기에 사용하던 코드인데 제거되지 않았었네요. 😅
불필요한 코드이므로 제거했습니다! 

💾 dcfcd15cf7f02947c6235238c92d08d9de5f2f22

GREENOVER commented on ProjectManager/ProjectManager/Sources/Views/TaskList/TaskListView.swift:
```diff
@@ -0,0 +1,190 @@
+        }
+    }
+//
+//    func test(of task: Task, to status: Status) -> [some View] {
```
주석 코드와 데드 코드들은 최대한 없애는것이 좋습니다!🙋🏻

minsson replied:
앗..! 가장 최근에 한 커밋 하나가 push되지 않은 상태였습니다..! ( 872bad7f6291f43771db6f2dcecdcf0f5b82c432 )
PR 보내기 전 마지막으로 커밋한 거였는데, 불필요한 주석을 제거하고, Swipe Action들 컬러를 바꾼 거였어요. 😅

GREENOVER commented on ProjectManager/ProjectManager/Sources/Views/TaskList/TaskListView.swift:
```diff
@@ -0,0 +1,190 @@
+
+private extension TaskListView {
+    
+    func swipeButtonForChangingStatus(of task: Task, to status: Status) -> some View {
```
메서드로 뷰를 반환하여 사용하는점 너무 좋네요!
그런데 중복 코드들이 너무 많은데 아래와 같이 표현해보는건 어떻게 생각하시나요?🙋🏻
물론 Status에서 color 연산 프로퍼티를 추가해줘야합니다.
```swift
func swipeButtonForChangingStatus(of task: Task, to status: Status) -> some View {
  return Button {
    withAnimation(.spring()) {
      tasksDataSource.transfer(selectedTask: task, to: status)
    }
  } label: {
    Label("Move to\n\(status.description)", systemImage: "circle")
  }
  .tint(status.color)
}
```

minsson replied:
와... 제가 왜 이 생각을 못 했을까요? 심지어 description은 이미 쓰고 있었는데...! 😳 
감사합니다!!

5e478ce1809bd5545c2774456429d6412ef27661 에서 아래 코드와 같이 바꿔주었어요!

```swift
    func swipeButtonForChangingStatus(of task: Task, to status: Status) -> some View {
        return Button {
            withAnimation(.spring()) {
                tasksDataSource.transfer(selectedTask: task, to: status)
            }
        } label: {
            Label("Move to\n\(status.description)", systemImage: status.image)
        }
        .tint(Color(status.color))
    }
```

각 status별로 color와 image가 다르므로 둘 다 연산프로퍼티를 사용했어요.

그런데 color의 경우 `.tint(status.color)`를 사용하려면 `Status`의 연산프로퍼티가 `Color` 타입을 반환해야 하고,
이를 위해서는 `Status` 파일에서 `import SwiftUI`를 해야 하더라고요.

이 작업 하나를 위해서 프레임워크를 가져오는 게 맞을까? 라는 고민이 들었어요.

그래서 `String` 타입을 반환하게 만들고, 사용하는 쪽에서 `.tint(status.color)` 대신에 `.tint(Color(status.color)`로 바꿔봤어요. 
그린은 어떻게 생각하시나요? 🙋‍♂️

minsson replied:
와... 제가 왜 이 생각을 못 했을까요? 심지어 description은 이미 쓰고 있었는데...! 😳
감사합니다!!

https://github.com/yagom-academy/ios-project-manager/commit/5e478ce1809bd5545c2774456429d6412ef27661 에서 아래 코드와 같이 바꿔주었어요!

    func swipeButtonForChangingStatus(of task: Task, to status: Status) -> some View {
        return Button {
            withAnimation(.spring()) {
                tasksDataSource.transfer(selectedTask: task, to: status)
            }
        } label: {
            Label("Move to\n\(status.description)", systemImage: status.image)
        }
        .tint(Color(status.color))
    }
각 status별로 color와 image가 다르므로 둘 다 연산프로퍼티를 사용했어요.

그런데 color의 경우 .tint(status.color)를 사용하려면 Status의 연산프로퍼티가 Color 타입을 반환해야 하고,
이를 위해서는 Status 파일에서 import SwiftUI를 해야 하더라고요.

이 작업 하나를 위해서 프레임워크를 가져오는 게 맞을까? 라는 고민이 들었어요.

그래서 String 타입을 반환하게 만들고, 사용하는 쪽에서 .tint(status.color) 대신에 .tint(Color(status.color)로 바꿔봤어요.

그린은 어떻게 생각하시나요? 🙋‍♂️

GREENOVER replied:
말씀하신 방향도 괜찮습니다~
그런데 그 방향이면 제 생각은 유지보수가 더 어려울것 같아요.
String을 정의해줄때 오탈자가 날 확률이 크다고 생각해요.
우리가 모든 컬러나 아이콘등의 에셋 String 값을 매번 찾아야하고 자동완성이 되지 않기에 실제적으로 에러가 나면 어디서 에러가 발생했는지 코드가 많아지면 찾기 어렵지 않을까요!?🤔

minsson replied:
😵 그린 말에 설득되었습니다...! 
불필요한 프레임워크 하나 줄이자고 유지보수 용이성을 희생하는 건 너무 손해인 것 같네요! 😲

`Status`에서 `SwiftUI`를 import 하고, `color` 연산프로퍼티가 `Color` 타입을 반환하도록 수정했습니다!
호출하는 부분은 아래와 같이 수정했습니다! 
더 깔끔해졌네요! 😁

```swift
    func swipeButtonForChangingStatus(of task: Task, to status: Status) -> some View {
        return Button {
            withAnimation(.spring()) {
                tasksDataSource.transfer(selectedTask: task, to: status)
            }
        } label: {
            Label("Move to\n\(status.description)", systemImage: status.image)
        }
        .tint(status.color)
    }
```


💾 302114370d405fb01373701d014304bf89ed0b92

GREENOVER commented on ProjectManager/ProjectManager/Sources/Views/TaskCell/TaskCellView.swift:
```diff
@@ -0,0 +1,57 @@
+                Text(task.dueDate, formatter: Date.formatter)
+                    .foregroundColor(isOver(task.dueDate) ? .red : .primary)
+            }
+            .padding(EdgeInsets(top: 16, leading: 16, bottom: 16, trailing: 16))
```
`.padding(16)`로 간단히 표현할수도 있습니다🙌

minsson replied:
앗, 그렇네요! 감사합니다!! 😁
`.padding(16)`으로 수정했습니다! 

💾 ef22f77fd94dbf19f361140d9252837f9c6b00e0

GREENOVER commented on ProjectManager/ProjectManager/Sources/Views/TaskCell/TaskCellView.swift:
```diff
@@ -0,0 +1,57 @@
+                    .foregroundColor(isOver(task.dueDate) ? .red : .primary)
+            }
+            .padding(EdgeInsets(top: 16, leading: 16, bottom: 16, trailing: 16))
+            .frame(maxWidth: .infinity, alignment: .leading)
```
infinity가 필요한 이유가 어떤거라 생각하시나요~?🙋🏻

minsson replied:
결론적으로, 현재 코드에서는 `.frame(maxWidth: .infinity, alignment: .leading)`를 쓰지 않고 `ZStack`의 이니셜라이저에서 `alignment`를 `.leading`으로 지정해주면 됩니다. 그 편이 더 명확하다고 생각해 수정했습니다.

💾 ada7826ace6b01a64bd87bd26854373d5e57164a

---

### 기존 코드에서 `.frame(maxWidth: .infinity, alignment: .leading)`가 필요하다고 생각한 이유

` .frame(maxWidth: .infinity, alignment: .leading)`는 각 Task Cell의 title, description, dueDate를 왼쪽 정렬하고자 작성한 코드입니다.

maxWidth에 `.infinity`를 주면 해당 뷰의 너비가 늘어날 수 있는 한 최대로 늘어납니다. `alignment`와 함께 정렬 용도로 자주 사용되는 편인 것 같습니다. 때문에 `Text`와 같이 이니셜라이저 자체에서 정렬을 지원하지 않는 뷰의 경우 튜토리얼에서도, 여러 기술블로그나 스택오버플로우에서도 이렇게 사용하는 것을 볼 수 있었습니다.

처음에는 말 그대로 `무한`이 되는 거라면 예상하지 못한 문제가 발생할 수도 있지 않을까 하는 의문이 들었습니다. 🤔

그런데 [공식문서](https://developer.apple.com/documentation/swiftui/view/frame(minwidth:idealwidth:maxwidth:minheight:idealheight:maxheight:alignment:))를 찾아보니 Discussion 쪽에 아래와 같은 문장이 있어 의문이 해소되었습니다. 😁

> `If a maximum constraint is specified and the size proposed for the frame by the parent is greater than the size of this view, the proposed size, clamped to that maximum.`

GREENOVER commented on ProjectManager/ProjectManager/Sources/Views/TaskCreating+TaskEditing/Protocol/TaskWritableView.swift:
```diff
@@ -0,0 +1,48 @@
+
+import SwiftUI
+
+protocol TaskWritableView { }
+
+extension TaskWritableView {
```
오호 프로토콜 구현 좋네요!

GREENOVER commented on ProjectManager/ProjectManager/Sources/Views/TaskCreating+TaskEditing/Protocol/TaskWritableView.swift:
```diff
@@ -0,0 +1,48 @@
+            TaskDescriptionView(description: description)
+                .frame(maxHeight: 500)
+                .padding(.top, 10)
+                .overlay(alignment: .topLeading) {
+                    if description.wrappedValue.isEmpty {
+                        Text("필요한 경우 상세설명을 적어주세요")
+                            .foregroundColor(.gray.opacity(0.5))
+                            .padding(.top, 20)
+                    }
+                }
```
뷰에서 placeHorder와 같은 구현들은 여기뿐만 아니라 또 많이 사용될 여지가 있기에 저는 아래와 같이 View를 extension하여 별도 공통적인 메서드로 만들어 사용하기도 합니다!
```
extension View {
...
public func placeholder<Content: View>(
  when shouldShow: Bool,
  alignment: Alignment = .leading,
  @ViewBuilder placeholder: () -> Content) -> some View {
    ZStack(alignment: alignment) {
      placeholder().opacity(shouldShow ? 1 : 0)
      self
    }
  }
}
```

minsson replied:
와................마법 같네요...! 😳 
그린은 역시 최고예요..........! 😲 
감사합니다!

아래 코드와 같이 수정했어요!

```swift
    .placeholder(when: description.wrappedValue.isEmpty, alignment: .topLeading) {
        Text("필요한 경우 상세설명을 적어주세요")
            .foregroundColor(.gray.opacity(0.5))
            .padding(.top, 20)
    }
```

코드가 간결해질 뿐만 아니라, `placeholder`라는 네이밍 덕분에 어떤 역할을 하는 코드인지 훨씬 더 명료해졌어요! 
앞으로 재사용될 여지가 있는 모디파이어들은 이런 식으로 미리 만들어 놓아야겠어요!! 감사합니다!

💾 22fdac5e70c1c3db803fb58ef4d63e80013de8d6

GREENOVER reviewed (changes requested):
@minsson 
SwiftUI를 처음 접하고 프로젝트에 적용하기가 정말 쉽지 않으셨을텐데 구조적으로도 그렇고 아주 잘 사용해주셨네요!
학습이 다소 오래 걸린만큼 더 빛을 발하는 프로젝트가 된것 같아요ㅎㅎ
정말 빈말이 아니고 SwiftUI를 잘 이해하고 사용하셨다고 생각합니다🙌
또 4주간 포기하지 않고 새로운것을 늘 학습하고 적용해나가는 민쏜의 열정에 저도 한 수 배워갑니다🙇🏻
4주간 야곰에서의 마지막 프로젝트 너무 고생 많으셨고 대미를 장식하기 위해 저도 최대한 도움이 되도록 리뷰를 드려보겠습니다~!
***
### 칭찬 드리고 싶은 부분
- 구조적으로 아주 명확하게 분리를 잘 해주셨어요 💯
- 가장 기본적인 네이밍에 대해 애매한곳도 없고 리뷰드릴것도 없을만큼 훌륭합니다 💯
- SwiftUI를 왜 써야되는지 그리고 어떻게 써야되는지 감을 잡으신것 같아요 💯
- 기본에 충실해 군더더기 없는 코드들이 돋보였습니다 💯
***
### 보완되면 좋을 부분에 대한 저의 의견
우선 전체적으로 기능 동작 등에 대해서는 전혀 이상이 없고 사용성에서 보완되면 좋을 부분에 대해 얘기해보려합니다!
- Task 생성 및 수정 Sheet에서 별도 cancel 버튼이 없어 도중에 나가고 싶다면 백그라운드 외부 영역을 클릭해 Sheet를 닫아줘야하는 불편함이 있어 이 부분에 대해 사용성이 개선되면 좋겠습니다🙋🏻
- 현재 Sheet가 뜨고 수정 버튼을 눌러 편집 모드가 활성화 되는것에 대한 어떠한 표기가 있으면 좋을것 같다 생각했습니다..!
명세에는 어긋나지 않지만 뭔가 수정을 누르고 편집을 안하고 바로 편집을 하는것이 자연스러운 사용자도 꽤 있기에 수정중인지 아니면 읽기전용중인지 등 보완되면 어떨까해요🙋🏻
- 에러가 발생할 수 있는 상황은 없을지 생각해봐도 좋은 포인트일것 같습니다🙋🏻
***
### 고민 및 궁금하신 부분에 대한 저의 의견
1. `그린이 보시기에는 어떤지, 구조적으로 어떤 점을 개선하면 좋을지 가르쳐주시면 감사하겠습니다. 우선 제가 애매하게 느껴졌던 부분은 코드 리뷰 쪽에 먼저 코멘트를 달아놓겠습니다.`
- 애매하게 느끼셨던 부분에 대한 아직 코멘트가 없어 확인은 해보지 못했지만 전체적으로 현재 구성해주신 구조 적절하다고 생각합니다 저는!
2. `앞으로 SwiftUI를 꾸준히 공부하고 사용하고 싶습니다. 앞으로 어떤 아키텍처를 더 공부해보면 좋을까요? 그린의 블로그에서도 TCA에 대한 이야기를 많이 다루고 있어, 한 번 공부해보고 싶은데 먼저 공부하거나 체득해야할 개념이 있을까 해서 방향성에 대해 여쭤봅니다.`
- SwiftUI를 하면 이런 아키텍쳐를 써야한다! 이런건 당연히 없지만 저는 개인적으로 MVI 아키텍쳐가 현재 핫하기도하며 꽤 앱 개발에 있어 적절한 아키텍쳐라고도 생각하고 있습니다. ReactorKit, TCA등이 이 MVI 아키텍쳐 기반이라고 볼 수 도 있구요.
그렇기에 MVI 아키텍쳐가 갖는 특징과 장점을 보면 선언형 UI와 어떻게 연계 할 수 있고 사용할 수 있구나 체감하실 수 있을거라 생각합니다..!
3. `저는 Environment Object의 단점을 아직 알아내지 못했는데, 혹시 그린은 어떻게 생각하시나요?`
- 정리를 잘해주셨고 마찬가지로 싱글턴이 가질 수 있는 단점과 유사할 수 있다고 생각해요.
제가 생각할때 가장 큰 경계해야할 포인트는 싱글턴 인스턴스가 다양한 데이터를 공유할 때 각 클래스간에서 해당 인스턴스 결합도가 높아지게 될것이고 이는 SOLID에서 OCP, 개방폐쇄원칙을 위반하게 된다고 생각합니다.
그렇게 되면 제가 생각할때는 테스트가 어려워질 수 밖에 없습니다.
안된다는것이 아니고 까다롭다는거로 생각됩니다.
단순히 테스트를 위해 Mock 객체를 생성해보시려면 왜 어려운지 느껴지실 수 있습니다.
또 결국 위와 동일한데 의존성을 분리해야되는것이 중요하다는것은 아실텐데 어디서든 접근할 수 있다는것 자체가 의존성이 아주 강하게 결합되어 있다는 소리이기에 이곳저곳 의존성을 만들게 됩니다.
또, 운영체제 측면으로 넘어간다면 멀티 스레드를 사용하는 다양한 환경에서 여러곳에서 동시 다발적으로 접근할 일도 있을테고 이에 대한 처리가 미숙하다면 race condition을 발생시킬 확률이 다분합니다.
4. `폴더 및 파일 구조`
- 저는 현재 폴더 및 파일 구조 아주 적절하게 분리되어 있다고 느꼈습니다👍
5. `SSOT 역할을 하는 TasksDataSource의 네이밍과 역할 분리가 잘 되어 있다고 볼 수 있을까요?`
- 네 요부분도 네이밍 및 역할 분리가 어느정도 잘 되어 있다고 생각해요👍
6. `Task 데이터 관리`
- 음 정말 스타일의 차이일것 같긴한데 저라면 `9b2246c` 커밋처럼도 충분히 이점이 있을것 같다고 생각하고 그렇게 했을것 같습니다.
todo, doing, done외에도 더 많은 상태가 들어올때 유지보수가 조금 더 용이할것 같고 filter를 걸어 status를 내부적으로 처리하는것이 현재 todoTask, doingTask, doneTask를 가지고 switch문에서 각 분기 처리되는 중복 코드들을 조금 더 줄여주는 방향의 코딩도 가능할것 같다고 생각해요!
물론 지금 구조도 너무 명확하고 충분히 이점이 있다고는 생각합니다🙋🏻
7. `각 방법에 대한 장단점과 그린의 선택 기준이 궁금합니다! 또, 이 방법 외에도 다른 방법이 있는지 궁금합니다.`
- 민쏜이 생각하시는 각 방법의 장단점은 무엇인가요?
그걸 먼저 말씀해주시면 제가 도움을 드릴 수 있을것 같아요🙋🏻
현재 민쏜이 구현하신 방법도 충분히 좋습니다.
이 부분도 정말 딱 이게 성능적으로 좋다 뭐 어떻다하는 정답이 없는 문제이긴하지만요.
저는 private view에 대해서는 개인적으로 별도 해당 파일에서 프라이빗한 View를 만들어 사용합니다!
8. ` Task 생성 뷰와 편집 뷰`
- 저는 이정도도 충분히 괜찮다고 생각하는데 조금 더 중복된걸 합치고 싶다면 하나의 Create, Edit을 할 수 있는 뷰로 통합해 내부적으로 수정, 생성이라는 상태값을 두어 분기 처리를 하면서 뷰를 그려줘도 괜찮을것 같다고 생각합니다.
음 그리고 프로토콜 관련 질문은 조금 더 확대해서 해석해보자면 어떻게 자연스럽게 기술들을 사용하는건지에 대한 질문으로 이해했습니다!
이는 현업이라고 그렇다 이런건 없을것 같고, 점점 민쏜도 학습하다보시면 하나씩 익숙해지고 눈에 들어올거라 생각하기에 별도 학습법이 존재하는지는 잘모르겠습니다!
개인적으로 꾸준히 하다보면 자연스러워지는것 이라고 생각합니다🙋🏻
답변이 좀 애매하긴한데... 제 가치관에서의 기술들 노하우는 꾸준한 학습이라고 생각해요!
9. `아래와 같이 구현하면 코드를 최소화할 수 있다고 생각했는데, 혹시 이런 식으로 구현하는 것도 괜찮을까요?`
- 네! 무척 좋습니다ㅎㅎ 잘해주셨다고 느꼈던 코드입니다👍
***
코드에서 불필요하다고 생각되는 부분이 한두군데 정도 있어 우선 Request Changes로 드리겠습니다!
언제든 재리뷰 요청해주셔도 됩니다🙌
다시 한번 4주간 너무 고생 많으셨어요 민쏜ㅎㅎ
이제 정말 수료가 코앞이군요... 마무리 잘하고 이제 현업에서 뵈요!
화이팅입니다🥳🥳🥳

minsson committed:
refactor: "TaskListView" 스와이프 버튼 메서드 관련 중복 코드 제거

- Status에 String 타입의 color, image 연산프로퍼티 추가
    - color의 경우 Color 타입을 반환하는 연산프로퍼티로 만들 수 있으나, String 타입으로 구현
        - 사용하는 쪽에서 Color(status.color)로 사용할 수 있으므로, import SwiftUI가 필요하지 않다고 생각함

minsson committed:
refactor: "TaskDataSource" 강제언래핑 if let 바인딩으로 변경

minsson committed:
refactor: "TaskDashboardView" 불필요한 프로퍼티 제거

- newTask를 만들고 관리하는 역할을 TaskCreatingView로 분리하게 되면서 더 이상 사용하지 않게된 프로퍼티
    - 구현 초기에는 TaskDashboardView에서 sheet를 띄우면서 newTask를 주입해주는 방식으로 구현했었음

minsson committed:
refactor: "TaskCellView" padding 이니셜라이저 더 간단하게 변경

minsson committed:
refactor: "TaskCellView" ZStack 정렬 코드 리팩토링

- ZStack을 정렬하는 상황이므로 기존 코드와 같이 사용할 이유가 없음
- 기존 코드는 Text 등을 정렬할 때 사용

minsson committed:
refactor: "TaskListView" 불필요한 ZStack 제거

minsson committed:
refactor: "AddingTaskButtonView" 재사용하지 않는 타입이므로 삭제

- "TaskDashboardView"에서만 쓰이는 버튼이므로 해당 타입 내에서 버튼을 생성하는 메서드로 구현함

minsson committed:
refactor: "View+Extensions" placeholder 모디파이어 구현

- placeholder는 다양한 곳에서 사용될 여지가 있으므로 View를 extension하여 공통적인 모디파이어로 구현
- 재사용의 편의성이 향상되며, 관련 코드의 역할이 더 명료해짐

- "TaskWritableView"
    - placeholder 모디파이어 적용
    - 기존 overlay를 통해 구현한 것 대비 코드가 간결해지고, 어떤 역할을 하는 코드인지 명료해짐

minsson committed:
refactor: "Status" 유지보수 용이성을 위한 color 연산프로퍼티 타입 변경

- String 타입에서 Color 타입으로 변경
- String일 경우 자동완성이 되지 않는 것과 관련 오탈자 리스크를 줄이기 위함

- "TaskListView"
    - 위에서 수정한 코드 적용

GREENOVER commented:
@minsson 
코멘트에 대해 전부 답변 달아주신것은 확인했습니다!
혹시 재리뷰가 필요하시면 우측 상단 리프레쉬 버튼을 눌러 요청해주시면 됩니다🙌
![스크린샷 2022-10-07 오전 3 05 13](https://user-images.githubusercontent.com/72292617/194386590-c3d824a0-b1b2-4cc7-a220-8f24df0a9f29.png)

minsson committed:
feat: "ErrorAlertManager" 에러 핸들링을 위한 유틸리티 타입 구현

- 타입 프로퍼티와 타입 메서드를 활용하여 전역에서 접근할 수 있도록 구현

- "LogicError"
    - 에러가 발생할만한 상황에 맞는 에러 타입 구현 및 표시할 메시지 지정
    - 현재 에러가 발생할 수도 부분은 수정 및 삭제로 보여 관련 에러 케이스만 구현

- "TasksDataSource"
    - guard let 바인딩 중 nil이 나오는 경우 ErrorAlertManager에게 요청에 에러를 처리하도록 구현

- "TaskEditingView", "TaskListView"
    - 에러가 발생할 수도 있는 구간이므로, tasksDataSource에서 에러 발생시 에러가 표시되도록 구현

minsson committed:
feat: "TaskEditingView" 버튼에 편집모드, 읽기전용모드 활성화 표시

- 기존 '수정' 버튼을 편집 가능 여부에 따라 '편집모드 활성화', '읽기전용모드 활성화'가 되도록 표시
- 프로젝트 요구사항에 따라 이렇게 구현했지만, 기존의 Task Cell 탭 후 별도의 '편집모드' 없이 수정하는 게 더 직관적인 UX라고 생각함

minsson committed:
feat: "TaskCreatingView", "TaskEditingView" 취소버튼추가

- Task 생성 및 수정 Sheet에서 별도 cancel 버튼이 없어, 취소 필요 시 Sheet 외부 영역을 탭해야 하는 불편함에 대한 리뷰어 피드백 반영

minsson commented:
그린, 안녕하세요! 😁

언제나 그렇듯 세심하고 배려 깊은 리뷰 감사 드립니다! ☺️

처음 SwiftUI로 프로젝트를 진행하면서 쉬울 거란 예상과 달리 무척 어려웠고, 특히 뷰 간 데이터 공유 측면에서 많이 혼란스러웠습니다. 그래도 나름대로 SwiftUI 문서도 읽고, 여러 튜토리얼도 해보면서 최선을 다해 구현하고 뿌듯했는데, 그린이 좋게 평가해주셔서 더더욱 뿌듯했습니다!

중간에 수정을 위한 TaskEditingView와 Task Cell 간 데이터 전달에서 오랫동안 막혀있을 때 그린이 정답을 주시지 않고 제 생각의 흐름을 쭉 들어주시고, 약간의 가이드만 해주신 점도 감사했습니다. 긴 시간 동안 답답하셨을텐데 생각의 흐름을 유도해주셔서, 저도 설명하면서 생각을 잘 정리할 수 있었습니다. 무엇보다도 `두, 세 걸음 앞까지 미리 생각하지 말고, 우선은 지금 내딛는 걸음부터 생각하라`고 말씀해주신 게 많은 도움이 되었습니다. 일정에 대한 압박감을 내려놓고, 차분히 생각해보니 그 전에는 보이지 않던 것들에 대한 실마리가 잡혔습니다. 앞으로도 마음이 급해질 때마다 그린이 말씀해주신 걸 되새겨 보려고 합니다.

캠프 수료와는 별개로 앞으로도 계속 연락할 예정이니 작별인사는 하지 않겠습니다! ☺️

## 그린이 말씀해주신 보완 지점에 대한 답변
>우선 전체적으로 기능 동작 등에 대해서는 전혀 이상이 없고 사용성에서 보완되면 좋을 부분에 대해 얘기해보려합니다!

### 1) Cancel 버튼
>Task 생성 및 수정 Sheet에서 별도 cancel 버튼이 없어 도중에 나가고 싶다면 백그라운드 외부 영역을 클릭해 Sheet를 닫아줘야하는 불편함이 있어 이 부분에 대해 사용성이 개선되면 좋겠습니다🙋🏻

아래와 같이 취소 버튼 구현했습니다. 

| 구현 화면  |
|:---:|
|<img src=https://i.imgur.com/3QSJekM.png width=50%>|

원래 개인적으로는 취소버튼이 필요하지 않다고 생각했었습니다. 화면의 다른 부분을 탭하여 취소하는 것도 괜찮고, 그보다도 Sheet를 쓸어내려 dismiss 하는 방법을 지향하고 싶었습니다. 

그런데 그린이 취소 버튼에 대한 피드백을 주셔서 잘 생각해보니, 여러 앱에서 취소 버튼을 기본적으로 제공하고 있으며, 또, 저희가 구현할 때도 기본 옵션 중 하나로 `.cancel` 등이 있다는 점이 떠올랐습니다. 이와 관련해 [HIG - Buttons](https://developer.apple.com/design/human-interface-guidelines/components/menus-and-actions/buttons/)를 보니 아래와 같이 시스템 버튼의 역할 중 하나로 `cancel`을 보여주고 있어 취소 버튼이 있는 게 바람직하다는 생각이 들었습니다. 왜 가능한한 제스쳐를 많이 쓰는 게 좋다고 생각했는지 저도 잘 모르겠네요. 😅


💾 fa378f965d7fde85a26220240a5353c5ddce9943

```plain
Role
A system button can have one of the following roles:
Normal. No specific meaning.
Primary. The button is the default button — the button people are most likely to choose.
Cancel. The button cancels the current action.
Destructive. The button performs an action that can result in data destruction.
```

### 2) 편집모드 활성화 여부 표시
>현재 Sheet가 뜨고 수정 버튼을 눌러 편집 모드가 활성화 되는것에 대한 어떠한 표기가 있으면 좋을것 같다 생각했습니다..!
>명세에는 어긋나지 않지만 뭔가 수정을 누르고 편집을 안하고 바로 편집을 하는것이 자연스러운 사용자도 꽤 있기에 수정중인지 아니면 읽기전용중인지 등 보완되면 어떨까해요🙋🏻

피드백 주신대로 아래와 같이 보완했습니다!
| 구현 화면 |
|:---:|
|<img src=https://i.imgur.com/fKekM11.png width=50%>|

말씀해주신 것처럼 명세의 요구사항에 따라 기존과 같이 구현했었습니다.

다만, 그린이 말씀해주신 것처럼 `별도의 수정 버튼 탭 없이 바로 편집을 하는 것이 자연스러운 사용자가 꽤 있다`는 점에 대해 완전히 공감합니다. 생산성 앱을 많이 써본 입장에서, 한번도 `수정`이라는 버튼을 따로 눌러본 기억이 없습니다. 요구사항을 보고 구현하면서 무척 어색하다고 느꼈던 부분입니다. 

만약 제가 이런 앱을 출시한다면 별도의 수정 및 저장 버튼 없이, 아무 때나 탭하여 수정할 수 있고, 저장 버튼 없이도 자동저장 되도록 구현할 것 같습니다. (물론 캠프의 과제가 상용 앱을 만드는 것이 아니라 구현을 통한 학습을 추구한다는 측면에서, 어떻게 편집 가능 여부를 Bool 타입으로 전환해가며 기능을 활성화/비활성화 할지에 대한 고민은 유의미했습니다.)

💾 6551db8f7b322249d6cef8befc9e34ff30e59451

### 3) 에러 처리
>에러가 발생할 수 있는 상황은 없을지 생각해봐도 좋은 포인트일것 같습니다🙋🏻

에러가 발생할 수 있는 상황을 가정하여 아래와 같이 구현해보았습니다!
|구현 화면 (수정사항 저장 중 에러) | 구현 화면 (Task 삭제 중 에러) |
|:---:|:---:|
|<img src=https://i.imgur.com/fxDg95m.png width=70%>|<img src=https://i.imgur.com/iL4WoxW.png width=70%>|

이 부분에 대한 고민이 정말정말 많았습니다.

테스트와 에러처리가 정말 중요하다는 점에 대해 200% 공감하면서도 아직도 에러 처리를 잘 하지 못한다는 점이 무척 아쉽습니다. 아직 [The Swift Programming Language - Error Handling](https://docs.swift.org/swift-book/LanguageGuide/ErrorHandling.html)의 수준을 벗어나지 못한 것 같습니다.

이번에는 새로운 시도를 해보고자 Error 핸들링을 위한 구조체를 만들고, 그 구조체의 타입메서드를 통해 에러를 처리해봤습니다. 

그런데, `Error` 프로토콜을 상속 받는 `LogicError`라는 타입을 만들었지만, `throws`를 하지 않고, 이에 따라 `do-catch` 문도 없습니다. 아직 에러처리가 무엇인지 정확히 이해하지 못했다는 생각도 듭니다. 지금은 그저 에러가 날 수 있는 상황에 Alert를 띄우는 것에 불과하며, 이에 따라 `LogicError`는 `Error` 프로토콜을 상속받을 이유가 없습니다.

[The Swift Programming Language - Error Handling](https://docs.swift.org/swift-book/LanguageGuide/ErrorHandling.html) 외에 더 좋은 레퍼런스를 찾아보았지만, 아직 여기서 크게 벗어나는 글은 찾지 못했습니다. 혹시 이 부분에 대해 어떤 방향으로 공부해보면 좋을지 조언해주실 수 있나요? 🙋🏻

💾 4d3e7e7364dcfb08a20e0ca23b1a3f1fa839e56b

GREENOVER commented:
고생 많으셨습니다 민쏜!
아주 더 좋은 코드가 되었네요ㅎㅎ 머지하겠습니다🙌
에러처리에 관해서는 더 크게 벗어나는 레퍼런스를 찾아야하는지 어떤 부분이 궁금하신지 궁금해요.
말씀하신 가이드 문서를 토대로 모든 개발자들이 레퍼를 썼을거라 해당 문서에서 에러처리가 무엇인지 4단계로 어떻게 처리하는지 조금 더 여러번 읽어보다보면 이해가 좀 더 되는 지점이 있을것이고 부가적으로 궁금하신 부분이 있을거 같아요.
현재 코드에서 에러가 날 수 있는 상황에 얼럿을 띄우는것에 불과하다고 하셨는데 그러한 로직 자체가 에러처리를 잘하시는것인데 어떤걸 더 하고 싶으신건지도 궁금합니다🙋🏻
민쏜이 생각하는 에러 핸들링이란 무엇일까요?
다시 한번 스텝2 고생하셨어요!
그리고 수료도 축하드립니다ㅎㅎ
